### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-type-assertion] False positives for "as const" assertions

### ISSUE THREAD

**Issue by ashvinsrivatsa** [typescript-eslint/typescript-eslint#8721]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=KYDwDg9gTgLgBAYwgOwM7wLYEMoGthRwC8cARNngaXFqoiugNwCwAUKJLHAJbIwEAzLAmBwAIlhhY4AbzZw4FfFABccGAE8wwCAMU5lLVgoBuWADYBXYGvRReAcyMBfNgMvIEMbikRRgksASUgAUAJRqwdJyxvRo8ABGtKIkMvqUqunKcM5GCv4wllDIsnAAdBVJqMAANHBmVjZk1LlszkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KAvkA&tokens=false

### Repro Code

```TypeScript
export const marker = "marker" as const;
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
  const base = { marker: marker };
  return { ...base, value: "" };
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
}
```

### Expected Result

The first line of this code should not report an error for the `as const` assertion, since this assertion is necessary - removing it is a TypeScript compilation error (see [TypeScript playground](https://www.typescriptlang.org/play?ssl=9&ssc=1&pln=9&pc=2#code/KYDwDg9gTgLgBAYwgOwM7wLYEMoGthRwC8cARNngaQNwBQokscAlsjAQGZYLBwAiWGFjgBvWnDgV8UAFxwYATzDAIHSTml0JANywAbAK7A56KKwDmdAL60OB5AhjMUiKMEHABQgBQBKOV7CYhJIaPAARliovCQi6pSy8dJwVlpwbjAGUMiicAB0BZHRADRwuobGZKQp1kA)).

### Actual Result

Starting in typescript-eslint@7.3 (presumably https://github.com/typescript-eslint/typescript-eslint/pull/8558), the first line of this code reports a @typescript-eslint/no-unnecessary-type-assertion error.

**Comment by bradzacher** [comment]
This seems like a bug in TS.
Without the `as const` the type of `marker` is `"marker"`. With it the type is the same.

This is evidenced by the fact that TS types your interface as requiring the property `marker: "marker"`, not `marker: string`.

The fact that TS widens the object type to have `marker: string` unless the variable declaration is specifically marked with `as const` seems like counter-intuitive and incorrect behaviour.

**Comment by klassm** [comment]
Another reproduction:

```ts
export const problemJsonTypeUriPrefix = 'https://errors.home-connect.com/' as const;
const errorBaseUri = `${problemJsonTypeUriPrefix}contentservices` as const;

export type ProblemJson = Record<string, unknown> & {
  type: `${typeof problemJsonTypeUriPrefix}${string}/${string}`;
};

export const ValidationError: ProblemJson = {
  type: `${errorBaseUri}/validation-error`,
};

```

This will remove both as consts. Usually `problemJsonTypeUriPrefix` comes from a library, so removing the const changes semantics and actually produces as compile error.

**Comment by JulienBara-360** [comment]
Hi @bradzacher,
I met the same problem today.
If it helps, rolling back from v7.4.0 to v6.21.0 removed the false positive.
In the meantime, I kept `typescript` on v5.4.3.

**Comment by kgregory** [comment]
Ran into this as well.  

It appears to have been introduced with [7.3.0](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v7.3.0), possibly with #8558 - Does not occur after rolling back to 7.2.0.

**Comment by kirkwaiblinger** [comment]
Hmmmm, note also that 
```diff
-export const marker = "marker" as const;
+export const marker = "marker";
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
- const base = { marker: marker };
+ const base = { marker: marker } as const;
  return { ...base, value: "" };
}
```
works... and makes a whole lot more sense anyway.  It seems _really odd_ that using `as const` in the first line would have an effect on the type of `base` granted `marker` has the same type whether or not `as const` is explicitly used.

**Comment by kirkwaiblinger** [comment]
```diff
-export const marker = "marker" as const;
+export const marker = "marker";
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
- const base = { marker: marker };
+ const base: { marker: typeof marker } = { marker: marker };
  return { ...base, value: "" };
}
```
works also.

As does 
```diff
-export const marker = "marker" as const;
+export const marker = "marker";
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
- const base = { marker: marker };
+ const base = { marker: marker as typeof marker };
  return { ...base, value: "" };
}
```

and, of course
```diff
-export const marker = "marker" as const;
+export const marker = "marker";
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
 const base = { marker: marker };
- return { ...base, value: "" };
+ return { marker, value: "" };
}
```

So it has something to do specifically with the widening behavior when creating a mutable object whose type is inferred. More I think about it, it really just seems surprising to me personally that the example in the issue report works in the first place.

**Comment by kirkwaiblinger** [comment]
> > works... and makes a whole lot more sense anyway.
> 
> In this case, yes, but that will affect all other attributes of the object. For example, if `base` were of a type with an array attribute, applying `as const` to `base` would type that array as `readonly`. (see [this playground](https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=KYDwDg9gTgLgBAYwgOwM7wLYEMoGthRwC8cARNngaQNwCwAUKJLHAJbIwEBmWCwcAESwwscAN4M4cCvigAuODACeYYBC7ScsuvSkA3LABsArsAXoo7AOY6pyYxgBGBVAvtOCAbQC6OgL4MXMbICDCsKIhQwMLAQiIAFACUCnGiErqIKOhwjlio-CRimpTyxbJwfrZwUTDGUMjicAB0Lbn5ADRwBiZmZKSd7s5QrnA%2BFXB5mWgw-kA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KAvkA&tokens=false))

@kgregory yeah, that's a fair point. and that irked me too. See https://github.com/typescript-eslint/typescript-eslint/issues/8721#issuecomment-2104880497 for 3 more examples that don't though 🤣

**Comment by kirkwaiblinger** [comment]
I also note that putting `const marker = 'marker' as 'marker'` also has the same impact in these examples as `const marker = 'marker' as const`.

I guess it seems like we're running into a case where a type assertion has broader effect than setting the type of the thing being asserted on.

- Maybe that's a bug in TS
- Maybe that's a intentional behavior in TS.
   - If so, _maybe_ it's possible to distinguish between the "necessary" and "unnecessary" cases
   - More likely, it's not possible/feasible to do so, and we'd need to choose between erring on the side of permitting unnecessary `as const` assertions or having this _sort of false positive_ edge case (wouldn't be the only one either, see https://github.com/typescript-eslint/typescript-eslint/issues/9007). 
   
Curious what other team members think

**Comment by kirkwaiblinger** [comment]
I've just learned that TS appears to have intentional behavior around the "freshness" of a `const` variable type (see https://github.com/microsoft/TypeScript/pull/52473#discussion_r1096104099), though it's not clear that it's a particularly external-facing concept. Would have to do some more research.

In any case, I think it explains what we're seeing here, though I'm not sure that it is obvious what we should actually do.

@bradzacher, thoughts?

**Comment by JoshuaKGoldberg** [comment]
@kirkwaiblinger I agree with you that we should get clarity on whether this is a TypeScript bug. +1 that it seems odd how having the `as const` on line 1 doesn't change the reported type of the `const marker`, but does change the type of the `const base = { marker: marker }`.

Marking as blocked on someone filing an issue on TypeScript.

**Comment by kirkwaiblinger** [comment]
> Marking as blocked on someone filing an issue on TypeScript.

Posed a question in the TS discord to start 👍

**Comment by sandersn** [comment]
This behaviour is intended in Typescript, to allow constant values to initialise mutable ones:

```ts
const start = 1001;
const max = 100000;
// many (thousands?) of lines later ...
for (let i = start; i < max; i = i + 1) {
  // did I just write a for loop?
  // is this a C program?
}
```

(From https://github.com/microsoft/TypeScript/wiki/Reference-Checker-Widening-Narrowing#literal-widening)

The best way to write this, in my opinion, is to put the `as const` in the mutable location:

```diff
-export const marker = "marker" as const;
+export const marker = "marker";
export interface Data {
  marker: typeof marker;
  value: string;
}
function createData(): Data {
- const base = { marker: marker };
+ const base = { marker: marker as const };
  return { ...base, value: "" };
}
```

Even better, if the program allows, is to use `"marker"` everywhere instead of extracting to a `const` variable. The compiler checks `"marker"` everywhere `Data` is used from here on out, so there shouldn't be a problem with typos which I guess is one reason for extracting to a variable.

**Comment by kirkwaiblinger** [comment]
@sandersn Many, many thanks for the excellent explanation and also the helpful reference to the typescript wiki! 🙏 

This info also helps us simplify the repro considerably:
```ts
const freshMarker = 'marker';
let mutableMarker = freshMarker;
const notAssignableDueToWidening: 'marker' = mutableMarker; // TS ERROR - can't assign string to 'marker'
```
vs 
```ts
const unfreshMarker = 'marker' as const; // TYPESCRIPT-ESLINT ERROR - no-unnecessary-type-assertion
let mutableMarker = unfreshMarker;
const notAssignableDueToWidening: 'marker' = mutableMarker; 
```
***

It's not immediately obvious to me what the best behavior for the rule would be in light of this info.

In theory, it might make sense to update the check to be: if the type before and after the assertion are the same _and_ it's not in a context in which freshness could be impacted, consider it an unnecessary type assertion. 

Seems like there might also be an argument that usually the freshness/unfreshness isn't necessary, so from a developer perspective this probably _is_ an unnecessary type assertion in most cases, albeit not one that behaves _identically_ after removing (and therefore which may need to be disabled when the unfresh behavior is genuinely desired). 

@typescript-eslint/triage-team

**Comment by kirkwaiblinger** [comment]
Note also that re-allowing 
```ts
const three = 3 as 3;
// or 
const four = 4 as const;
```
would basically amount to rethinking https://github.com/typescript-eslint/typescript-eslint/issues/4485 and https://github.com/typescript-eslint/typescript-eslint/issues/8458 (as noted by @ashvinsrivatsa when pointing out that this presumably changed with https://github.com/typescript-eslint/typescript-eslint/pull/8558)

**Comment by kirkwaiblinger** [comment]
What's the actual resolution here for which we're accepting PRs? @JoshuaKGoldberg 

Seems like for type assertions which only affect the freshness of types, which presumably is just `const x = "some primitive literal" as const` or `const x = "some primitive literal" as "some primitive literal"`, we can 
a) continue to report it, but have a special case error message, and document this edge case on the rule page. Maybe also convert the autofix to a suggestion for this special case, in order to give the user a chance to choose whether they want to disable the report or remove the assertion.
b) never report it (since it's not truly truly useless), and document this edge case
c) have an option for the user to select between a) and b)
d) some opinionated and confusing mix of the above, for example allowing `3 as const` but not `3 as 3`.

My head is at tentatively preferring option a).

**Comment by JoshuaKGoldberg** [comment]
Agreed. I don't love potentially falsely erroring on a case users are likely to hit, but this is a somewhat niche edge case and pretty easy to work around. +1 to option 1.

**Comment by bradzacher** [comment]
I'd say option 1 is best too - we don't want to skip reporting cos otherwise most of the rule's reports would go away.

We should be able to auto detect this case using scope analysis.

**Comment by LukeNotable** [comment]
My perspective here: We have a codebase with ~300 scattered cases of `export const FOO = 'foo' as const;`, which all get flagged by this rule. The auto-fix removes `as const` and results in ~25 scattered TypeScript errors. We currently have the rule disabled because of this, with a comment linked to this issue. I'd love to enable it (it's a useful rule!) without introducing ~300 `eslint-disable` directives. Remember, these are _false_ positives, since removing `as const` risks breaking any importing code. In fact, even for true positives where  `const ... as const` _can_ be safely changed as the code stands, I don't necessarily want to do so, since it isn't the same and can matter later on. 

I'd rather just disable this specific aspect of the rule (option 2) or at least have an option to do so (option 3), i.e., to always allow type assertions that affect freshness.

**Comment by JoshuaKGoldberg** [comment]
👍 that's helpful context, thanks @LukeNotable. I'd be in favor of option 2 or 3, adding an option to the rule.

**Comment by bradzacher** [comment]
> 300 scattered cases...
> 25 TypeScript errors...

So the rule has a pretty good false-positive rate, right? Assuming each error is caused by one false positive - it's correct 92% of the time.

So long as we remove the fixer for this case - that would mean that the rule would report and be correct 92% of the time and someone could manually fix it, and the remaining 8% of the time someone could suppress it.

1 still seems like the correct option, IMO

**Comment by LukeNotable** [comment]
> So long as we remove the fixer for this case - that would mean that the rule would report and be correct 92% of the time and someone could manually fix it, and the remaining 8% of the time someone could suppress it.

It would mean all 300 cases are reported by the rule. Removing the fixer would just make it harder _if_ we want to react to that by removing all (or most of) the `as const`. However, we probably don't, because (1) it would lead to inconsistency where some cases have `as const` (with an `eslint-disable`) and some don't, just to satisfy the linter, and (2) it partly undermines what the `as const` is there for in the first place. That is, the rule is only "correct" in many cases in the sense that it doesn't currently result in TypeScript errors, but it's still wrong in the sense that the `as const` is not wholly superfluous.

**Comment by dmurvihill** [comment]
>  the rule would report and be correct 92% of the time

is a terrible false positive rate.

I'd suggest simply allowing `as const`, if applied to a literal that can be widened.

**Comment by Maxim-Mazurok** [comment]
Hey guys, would prefer option 3 - never report it;
Since it does make a difference in TypeScript, and is a pretty popular feature of the language (even though more on the advanced side).

Here's another example where it makes a difference:

![image](https://github.com/user-attachments/assets/b7c388ec-8c32-4033-bc7a-c9958f32334a)

![image](https://github.com/user-attachments/assets/82274d70-d67d-41c0-888e-6cc841bc4c25)

Please at least provide an option for advanced TS users to allow `as const` case. Of course it would be awesome if we could auto-detect type widening possibilities, but probably isn't worth the effort on linter side.

**Comment by JoshuaKGoldberg** [comment]
Coming back to this, I'm still in the camp of making the rule lenient by default and adding an opt-in off-by-default option to be more strict. Two reasons why:

* The fact that TypeScript's behavior is so mysterious as to make us need to check in with the team hints to me that this will be a big confusing pain point if we enable it by default
* The `as const` practice in question is a common use case that people need - especially as evidenced by the big pushback here.

We could always turn the option on in `strictTypeChecked` - though IMO that should be a separate followup issue, to not block this one.

Proposal: let's add this option as off-by-default for our current major version (v8), and treat turning it on by default as a separate followup issue?

**Comment by kirkwaiblinger** [comment]
+1, I've been swayed based on community feedback and reasons articulated by @JoshuaKGoldberg

**Comment by JoshuaKGoldberg** [comment]
🚀 accepting PRs!

**Comment by KuSh** [comment]
Just encountered that on production code during eslint / tslint upgrade.
The autofix completely killed my 100k LOC project and unfortunately isn't disableable [atm](https://github.com/eslint/rfcs/pull/125)

Modifying a little bit the [playground link](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=C4TwDgpgBAsghgJwNYQVAvFA5AW0ShLAWAChSIAPMAewWCgGNqA7AZ3uqQygCI9lUPANzkqtekzb0k1bn3yCocVoxbsRJSjTqqpUHJwBcsBWkzyBCYaO0S19AEZwAJsYsEec-h403xUAEtmYFQAMzgGaAAROGA4KABvUih9U2N4Sw0UgDc4ABsAVwhjdgQggHMNAF9SUlCC5gZgAJZGBAhYiAB5JAAKAEpjGLjE5N12KCdWaEwE1MtjGSgqrKh24AKEZkSoADp9qYgAGihcwuLeTxXSGrISesbm1oZ2zoBpagGh2PikkhTJBNDtw5t5UMZOMtVutNts5vtdocTmcim4rtVavcGk0WtsXh0QjBOABGL5QYa-MaAxzKGY7MEIYwGLjXf5rCAbLY7BFI075VGXKE3TEPHHPV6EzgAJjJFNGbOpk1pIPmBGMTmcQrZMK58IOtOR-IuPHRNyAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RIAGnDY8E8lRr1kjFq2LREAMzlTktMkuT4VkNRsi6AviG9A&tsconfig=N4KAvkA&tokens=false) there's no easy fix when the Marker type isn't exposed simply (as in a string union deep inside an interface for example) and the rule is even in contradiction with prefer-as-const which is a recommended rule

**Comment by controversial** [comment]
Would it be possible for `typescript-eslint` to detect whether or not a constant variable is later used as part of type inference for another variable (i.e. a `let` or an object declaration) in order to decide whether or not an `as const` assertion is unnecessary?
